mock/py/mockbuild/backend.py

```diff
@@ -583,14 +584,24 @@ def buildsrpm(self, spec, sources, timeout, follow_links):
                 # Resolve any symlinks
                 sources = os.path.realpath(sources)
                 if os.path.isdir(sources):
-                    util.rmtree(self.buildroot.make_chroot_path(self.buildroot.builddir, "SOURCES"))
+                    host_chroot_sources = self.buildroot.make_chroot_path(
+                        self.buildroot.builddir, "SOURCES")
+
+                    util.rmtree(host_chroot_sources)
                     shutil.copytree(sources,
-                                    self.buildroot.make_chroot_path(self.buildroot.builddir, "SOURCES"),
+                                    host_chroot_sources,
                                     symlinks=(not follow_links))
                 else:
-                    shutil.copy(sources, self.buildroot.make_chroot_path(self.buildroot.builddir, "SOURCES"))
+                    host_chroot_sources = self.buildroot.make_chroot_path(
+                        self.buildroot.builddir, "SOURCES", os.path.basename(sources))
```

Comment:
I'm missing something here.  Can you document why `SOURCES` and now you need one more field `os.path.basename(sources)`?

Reply:
I think this else branch is for case when the `sources` is a tarball. Then we want to maintain the tarball name. `shutil.copy` in that case would also maintain the tarball name.

Reply:
> I think this else branch is for case when the sources is a tarball.

Yes.

> Then we want to maintain the tarball name. shutil.copy in that case
> would also maintain the tarball name.

I see, `shutil.copy('/file', '/dir')` is equivalent to
`shutil.copy('/file', '/dir/file')`, and you want to pass down tarball
name to the `pre_srpm_build`.

Why to do that?  It means that every `pre_srpm_build` needs to
check whether the path points to filename or directory because it is
uncertain.

Could we make it consistent, and always pass down to path to sources
directory?  If not, can you document the reasons?

Reply:
Well, i think it depends on what is being passed to `--sources <param>`. Whether a file or a subdirectory. I can for sure document this but can you point me where, please?

Reply:
Well, I mean ... why do you want to let `pre_srpm_build` depend on the
argument that way when we can design more clear API?  I don't see any
benefits, only negatives (I'd prefer passing down the path to SOURCES dir
always).

But if there _are_ some benefits or needs which are not really obvious,
we have to document so we don't have to ask this question again.

Reply:
I am not really sure if there is a significant benefit. A potential benefit could be that if user specifies a specific tarball and there is an srpm preprocessing plugin, which does certain preprocessing on tarballs (i.e. license checks) or that reads headers from the tarball to use them for spec preprocessing, then it is handy to know what tarball was specified on command-line by user because mock afaik doesn't really cleanup SOURCES so there might be more tarballs for some other packages too at this point. By not passing the source file user specified explicitly, we could therefore potentially lose some information. I would probably prefer just a comment in documentation of the hooks (if there is any) but let me know what you think and if this is ok with you. rpkg_preprocessor currently doesn't need that use-case when a user specifies `--sources <file>`.

Reply:
Thanks for the response,

> By not passing the source file user specified explicitly, we could
> therefore potentially lose some information.

When anyone ever needed the additional info, I think we could add
additional argument, --source argument could make a note in buildroot
object, or something like that.

>  rpkg_preprocessor currently doesn't need that use-case when a user
>  specifies --sources <file>.

Cool, can you please make the API simpler then, and always pass the source
directory path down to the hook?